### PR TITLE
New atm features rebased

### PR DESCRIPTION
**Pull request by kaeluka**
## This PR

This takes the `new-atm-features` branch (but excludes a few experimental commits near its top), rebases it on top of current `main`, and removes the obsolete features that have been replaced.

### CODE REVIEW

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll [hunk 1]
```diff
@@ -303,3 +207,580 @@
...
+  override string getName() { result = "contextSurroundingFunctionParameters" }
+
+  Function getRelevantFunction(DataFlow::Node endpoint) {
+    result = endpoint.asExpr().getEnclosingFunction*()
+  }
+
+  override string getValue(DataFlow::Node endpoint) {
+    result =
+      concat(string functionParameterLine, Function f |
+        f = getRelevantFunction(endpoint) and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll:337]: done locally

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll [hunk 2]
```diff
@@ -303,3 +207,580 @@
...
+   * function g(x, y, z) {
+   *   function h(u, v) { ... }
+   *   ...
+   * }
+   */
+  string getFunctionInterfacesForFile(File file) {
+    result =
+      concat(Function func, string line |
+        func.getFile() = file and
+        exists(func.getName()) and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll:547]: done locally

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/debug/CompareFeatures.ql [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+predicate isNone(string value) { value = "" }
+
+// query for comparing feature values
+from
+  DataFlow::Node endpoint, EndpointFeature feature1, EndpointFeature feature2, string featureValue1,
+  string featureValue2
+where
+  feature1 instanceof Input_ArgumentIndex and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/debug/CompareFeatures.ql:16]: done locally

File: javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/debug/CompareFeatures.ql [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  feature2 instanceof ArgumentIndex and
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/debug/CompareFeatures.ql:17]: done locally

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll [hunk 5]
```diff
@@ -303,3 +207,580 @@
...
+    result = SyntacticUtilities::getFunctionInterfacesForFile(endpoint.getFile())
+  }
+}
+
+/**
+ * Syntactic utilities for feature value computation.
+ */
+private module SyntacticUtilities {
+  bindingset[start, end]
+  string renderStringConcatOperands(DataFlow::Node root, int start, int end) {
```
[the remark this answers is not in the document]
Comment (kaeluka) [reply javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll:475]: done locally

**Comment by kaeluka** [comment]
Before we can merge this, we need to (at least) train/evaluate/release the models, and add a commit here that makes the branch depend on the new model.

**Comment by kaeluka** [comment]
I just requested a review from @henrymercer - this is a pretty big change set. This is the branch we're aiming to release, as discussed yesterday evening.

We've been using most of it for many months now, which IMO means it'd be most important to take a look at the changes I've done today.

I've spotted one minor bug in one of the features when doing this, but I'd rather keep everything as stable as possible to minimize changes as much as possible.

File: javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll [hunk 6]
```diff
@@ -16,102 +16,8 @@ private import FunctionBodyFeatures as FunctionBodyFeatures
...
-    // User.findOne(ENDPOINT);
-    // ```
-    //
-    // would have a callee API name of `mongoose`.
-    featureName = "calleeApiName" and
-    exists(API::Node apiNode |
-      AccessPaths::accessPaths(apiNode, false, _, result) and call = apiNode.getInducingNode()
-    )
-  )
+  exists(EndpointFeature f | f.getName() = featureName and result = f.getValue(endpoint)) and
```
Comment (henrymercer) [inline javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointFeatures.qll:19]: Great decision to turn this into a class.

**Review by henrymercer — COMMENTED** [review]
A bunch of comments around the definition of the new features.

> Before we can merge this, we need to (at least) train/evaluate/release the models, and add a commit here that makes the branch depend on the new model.  No comments on the tests — those look great.

In addition to what you state here, we should also run the security-extended vs security-extended-plus-atm experiment to confirm that this change keeps us within our runtime performance budget.

**Comment by kaeluka** [comment]
Experiment came back, 34% slowdown. @tiferet, is this a blocker?

**Comment by tiferet** [comment]
> Experiment came back, 34% slowdown. @tiferet, is this a blocker?

I think that's a question for @henrymercer: is this still within the acceptable time cutoff relative to a run without boosted queries?

**Comment by henrymercer** [comment]
The KR is 25%.  I think we should spend some time optimizing the new features to keep us within this performance budget.

**Comment by tiferet** [comment]
> Experiment came back, 34% slowdown. @tiferet, is this a blocker?

Is this a 34% slowdown relative to the unboosted version or relative to the current version of ATM?

Also, @kaeluka do you have ideas for how to optimize this? We should do that before proceeding with the new data generation, because we want to generate data from the final version of the PR exactly as we're going to merge it.

**Comment by kaeluka** [comment]
> Is this a 34% slowdown relative to the unboosted version or relative to the current version of ATM?

Relative to the unboosted version :)

> Also, @kaeluka do you have ideas for how to optimize this? We should do that before proceeding with the new data generation, because we want to generate data from the final version of the PR exactly as we're going to merge it.

Optimizing CodeQL performance is one area where I know next to nothing :( the best I can do is to look for "obvious" mistakes, but in my experience the obvious ones I'm able to recognize are also caught by QL-for-QL (and QL-for-QL is coming up with nothing for this PR). I'd be very eager to help out on the optimization, but I think I can't drive this.

**Review by henrymercer — APPROVED** [review]
LGTM!  A small point of feedback: for future PRs, particularly those with big diffs, please merge in main rather than force pushing if possible once someone's started a review.  That makes it possible to just review the things that have changed when giving a follow up review.  Thanks!

**Pull request merged**